File: product-processor-service/src/main/java/com/example/product_processor_service/service/common/kafka/template/router/KafkaTemplateRouterManager.java
package com.example.product_processor_service.service.common.kafka.template.router;

import com.example.product_processor_service.exception.NotFoundException;
import org.springframework.kafka.core.KafkaTemplate;
import org.springframework.stereotype.Component;

import java.io.Serializable;
import java.util.HashMap;
import java.util.Map;
import java.util.Optional;

@Component
public class KafkaTemplateRouterManager implements KafkaTemplateRouter {
    private final Map<Class<?>, KafkaTemplate<String, ?>> kafkaTemplateMap;

    public KafkaTemplateRouterManager(KafkaTemplate<String, String> stringKafkaTemplate) {
        kafkaTemplateMap = new HashMap<>();
        kafkaTemplateMap.put(String.class, stringKafkaTemplate);
    }

    @Override
    public <V extends Serializable> KafkaTemplate<String, V> route(Class<V> clazz) {
        return (KafkaTemplate<String, V>) Optional.ofNullable(kafkaTemplateMap.get(clazz))
                .orElseThrow(() -> new NotFoundException("unknown kafka template"));
    }
}

File: product-manager-service/src/main/kotlin/com/example/product_manager_service/service/kafka/template/router/KafkaTemplateRouterManager.kt
package com.example.product_manager_service.service.kafka.template.router

import com.example.product_manager_service.model.product.dto.ProductPublisherDto
import org.springframework.kafka.core.KafkaTemplate
import org.springframework.stereotype.Service
import java.io.Serializable
import kotlin.reflect.KClass

@Service
class KafkaTemplateRouterManager(productPublisherKafkaTemplate: KafkaTemplate<*, ProductPublisherDto>) :
    KafkaTemplateRouter {
    private val routerMap = mutableMapOf<KClass<*>, KafkaTemplate<*, *>>()

    init {
        routerMap[ProductPublisherDto::class] = productPublisherKafkaTemplate
    }

    override fun <T : Serializable> route(valueClass: KClass<T>): KafkaTemplate<String, T> =
        routerMap[valueClass] as? KafkaTemplate<String, T>
            ?: throw IllegalArgumentException("unknown kafka template for: $valueClass")
}